server/server.go

```diff
@@ -32,55 +62,170 @@ func StartHttpServer(opts *options.Options) error {
 	return nil
 }
 
-// Check that we can open a TPC connection to all the ports in opts.Ports
-func checkTcpPorts(opts *options.Options) *httpResponse {
+func checkHealthChecks(opts *options.Options) *httpResponse {
 	logger := opts.Logger
 	logger.Infof("Received inbound request. Beginning health checks...")
 
-	allPortsValid := true
-
+	// initialize failedChecks to 0, used as atomic counter for goroutines below
+	var failedChecks uint64
 	var waitGroup = sync.WaitGroup{}
 
-	for _, port := range opts.Ports {
+	for _, check := range opts.Checks {
 		waitGroup.Add(1)
-		go func(port int) {
-			err := attemptTcpConnection(port, opts)
+		go func(check options.Check) {
+			defer waitGroup.Done()
+			err := check.DoCheck(opts)
 			if err != nil {
-				logger.Warnf("TCP connection to port %d FAILED: %s", port, err)
-				allPortsValid = false
-			} else {
-				logger.Infof("TCP connection to port %d successful", port)
+				logger.Warnf(err.Error())
+				atomic.AddUint64(&failedChecks, 1)
 			}
-
-			waitGroup.Done()
-		}(port)
+		}(check)
 	}
 
 	waitGroup.Wait()
 
-	if allPortsValid {
-		logger.Infof("All health checks passed. Returning HTTP 200 response.\n")
-		return &httpResponse{ StatusCode: http.StatusOK, Body: "OK" }
-	} else {
+	failedChecksFinal := atomic.LoadUint64(&failedChecks)
+	if failedChecksFinal > 0 {
 		logger.Infof("At least one health check failed. Returning HTTP 504 response.\n")
-		return &httpResponse{ StatusCode: http.StatusGatewayTimeout, Body: "At least one health check failed" }
+		return &httpResponse{StatusCode: http.StatusGatewayTimeout, Body: "At least one health check failed"}
```

Comment:
Future TODO: it would be nice to list the health checks that failed or succeeded in the body of the HTTP response. To do this, we'd probably have the `goroutines` write their results (error or success) to a channel and then read and return all the data from the channel here. That way, no need for an atomic int. No need to do now though.